language/move-core/types/src/fs.rs

```diff
@@ -0,0 +1,535 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{de, export::Formatter, Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    cell::RefCell,
+    cmp::Ordering,
+    collections::HashMap,
+    ffi::OsString,
+    fmt, fs,
+    fs::{File, OpenOptions},
+    io::{Error, ErrorKind, Read, Write},
+    path::{Path, PathBuf},
+};
+
+thread_local! {
+     static DESCRIPTORS: RefCell<Option<Descriptors>> = RefCell::new(None);
+}
```

Comment:
Before I go too deep into the PR. I'd like to hammer out some details in this file. 

The reason I didn't initially do something like this is that at some point in the future, the Move source language compiler will want to be parallel/multi-threaded. At that time, I was worried any such implementation for interned file strings would fall apart. 

How will this work if the Move source language were to become multithreaded?

Reply:
Now the implementation cannot work in a multithreaded environment. But it is quite easy to fix,
we can do something like that.

```
static DESCRIPTORS: Lazy<RwLock<Option<Descriptors>>> = Lazy::new(|| RwLock::new(None));

#[derive(Debug)]
struct Descriptors {
    /// Context counter.
    fs_counter: usize,
    /// Path table.
    paths: Vec<String>,
}
```
The table will be created together with the context at the beginning of compilation and destroyed at the end.
This will avoid memory leaks.
This is the main goal of our PR.
I would like to discuss ways to solve this problem.